File: 19th/코딩테스트 심화반/week8/1800번_인터넷 설치.cpp
#include <iostream>
#include <queue>
#include <vector>

using namespace std;

vector<pair<int, int>> adj[1001];
int cost[1001][1001]; // 노드 번호, 공짜로 연결해준 인터넷 선의 수 

int N, P, K;

void dijkstra(int n) {
    for (int i = 1; i <= N; i++) fill(cost[i], cost[i] + N, 123456789); // 모든 노드까지의 거리를 inf로 초기화

    priority_queue < pair<int, pair<int, int>>, vector < pair<int, pair<int, int>>>, greater<pair<int, pair<int, int>>>> pq;
    pq.push({ 0, {n, 0} }); // 원장선생님이 내는 비용의 최솟값, 노드 번호, 공짜로 연결해준 인터넷 선의 수
    cost[n][0] = 0;

    while (!pq.empty()) {
        int c = pq.top().first; // 원장선생님이 내는 비용의 최솟값
        int x = pq.top().second.first; // 노드 번호
        int free_cnt = pq.top().second.second; // 공짜로 연결해준 인터넷 선의 수
        pq.pop();

        if (cost[x][free_cnt] < c) continue; // 현재 노드까지 이동한 비용의 최솟값보다 클 경우 continue

        for (int i = 0; i < adj[x].size(); i++) {
            int nx = adj[x][i].first;
            int nc = adj[x][i].second;

            // 만약 다음 인터넷을 연결할 때 원장선생님이 내는 비용의 최솟값보다 큰 비용이 들어갈 경우 아래의 두 가지로 분기
            // 1. 공짜로 연결
            // 2. 원장선생님이 내는 비용의 최솟값을 갱신
            if (c < nc) { 
                if (free_cnt < K) { // 공짜로 연결해준 인터넷 선의 수가 K개보다 작을 경우
                    // 공짜로 연결시키는 경우
                    if (cost[nx][free_cnt + 1] > c) {
                        pq.push({ c, {nx, free_cnt + 1} });
                        cost[nx][free_cnt + 1] = c;
                    }
                }
                // 원장선생님이 내는 비용의 최솟값을 갱신하는 경우
                if (cost[nx][free_cnt] > nc) {
                    pq.push({ nc, {nx, free_cnt} });
                    cost[nx][free_cnt] = nc;
                }
            }
            // 만약 다음 인터넷을 연결할 때 원장선생님이 내는 비용의 최솟값보다 적은 비용이 들어갈 경우
            else {
                if (cost[nx][free_cnt] > c) {
                    pq.push({ c, {nx, free_cnt} });
                    cost[nx][free_cnt] = c;
                }
            }
        }

    }
}

int main() {
    cin >> N >> P >> K;

    int a, b, c;
    for (int i = 0; i < P; i++) {
        cin >> a >> b >> c;

        adj[a].push_back({ b, c });
        adj[b].push_back({ a, c });
    }

    dijkstra(1);

    int ans = 123456789;
    for (int i = 0; i < N; i++) ans = min(ans, cost[N][i]); // N번 노드까지 도달한 경우 중 K개보다 작은 수의 인터넷 선을 공짜로 연결한 경우도 고려

    if (ans == 123456789) cout << -1; // N번 노드까지 도달하지 못한 경우
    else cout << ans;
}

File: 19th/코딩테스트 심화반/week8/1800번_인터넷 설치.py
import heapq

n, p, k = map(int, input().split())

adj = [[] for _ in range(n + 1)]
for _ in range(p):
    a, b, c = map(int, input().split())
    adj[a].append((b, c))
    adj[b].append((a, c))

INF = int(1e9)
# distance[x][f] = x번 노드에 f개의 무료 케이블을 사용해서 도달했을 때 최소 비용
distance = [[INF] * (n + 1) for _ in range(n + 1)]

def dijkstra(start):
    global distance
    q = []
    heapq.heappush(q, (0, (start, 0)))  # (현재 비용, (노드, 무료 사용 횟수))
    distance[start][0] = 0

    while q:
        cost, (x, free_cnt) = heapq.heappop(q)

        if distance[x][free_cnt] < cost:
            continue

        for nx, nc in adj[x]:

            # 현재 간선 비용이 더 크다면, 무료 케이블 사용 고려
            if cost < nc:
                # 무료 케이블 아직 쓸 수 있으면
                if free_cnt < k:
                    if distance[nx][free_cnt + 1] > cost:
                        heapq.heappush(q, (cost, (nx, free_cnt + 1)))
                        distance[nx][free_cnt + 1] = cost
                
                # 무료 안 쓰고 그냥 해당 간선 비용 사용
                if distance[nx][free_cnt] > nc:
                    heapq.heappush(q, (nc, (nx, free_cnt)))
                    distance[nx][free_cnt] = nc
            
            # 현재 경로 비용이 더 작거나 같은 경우
            else:
                if distance[nx][free_cnt] > cost:
                    heapq.heappush(q, (cost, (nx, free_cnt)))
                    distance[nx][free_cnt] = cost

dijkstra(1)

# N번 노드에 도달 가능한 모든 경우 중 최소 비용 선택
ans = INF
for i in range(n):
    ans = min(ans, distance[n][i])

if ans == INF:
    print(-1)
else:
    print(ans)